Add `AnnotatedOperation.params` and fix some control issues

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Adds support for `AnnotatedOperation.params` and, leveraging that, fixes some issues when controlling standard gates that are parameterized. Closes #10311, #10697 and #12135.

Since this PR also changes `AnnotatedOperation` it's probably too large for a backport?

### Details and comments

* `AnnotatedOperation.params` is a proxy to to `base_op.params`, if it exists. If the `base_op` has no `params`, returns an empty list (as discussed with the team offline).
* The `annotated` argument in `Gate.control` now is `None` per default, which allows to set it to `True` for gates that cannot be eagerly synthesized. A test for each gate that failed previously is added.
* The `S/Sdg` gates didn't return `CS/CSdg` gates for a single control, that's also been fixed.

Cryoris committed:
AnnotatedOp.params support and Gate.control fix

Cryoris committed:
add reno

Cryoris committed:
lint

Cryoris committed:
update reno

alexanderivrii commented on releasenotes/notes/annotated-params-116288d5628f7ee8.yaml:
```diff
@@ -0,0 +1,23 @@
+    The controlled version of some gates (e.g. :class:`.RXXGate` or
+    :class:`.RYGate` for more than 1 control) cannot be synthesized if
+    they contain unbound parameters. Previously, this raised an error, but
+    now we create an :class:`.AnnotatedOperation` as placeholder. This
```
Maybe say something like: "Previously, this raised an error when such gates were added to the circuit, but ..."

alexanderivrii commented on qiskit/circuit/annotated_operation.py:
```diff
@@ -219,6 +220,29 @@ def power(self, exponent: float, annotated: bool = False):
+            self.base_op.params = value
+        else:
+            raise NotImplementedError(
+                f"Cannot set attribute ``params`` on the base operation {self.base_op}."
```
I am wondering if there might be a better error message type than `NotImplementedError`.

Cryoris replied:
Maybe `AttributeError`?

ElePT replied:
AttributeError sounds about right:

```suggestion
            raise AttributeError(
                f"Cannot set attribute ``params`` on the base operation {self.base_op}."
```

alexanderivrii commented on qiskit/circuit/gate.py:
```diff
@@ -118,16 +117,16 @@ def control(
-            annotated: indicates whether the controlled gate can be implemented
-                as an annotated gate.
+            annotated: indicates whether the controlled gate is implemented
+                as an annotated gate. If ``None``, this is handled as ``False``.
```
I have not looked through all of the PR yet, but I am wondering if there is ever a difference between `annotated=False` and `annotated=None`.

Cryoris replied:
Yes there is: for all parameterized gates that cannot be synthesized until the parameters are bound. Some examples are: 
- RX or U with more than 1 control
- RXX with any number of controls

alexanderivrii commented on qiskit/circuit/library/standard_gates/multi_control_rotation_gates.py:
```diff
@@ -383,6 +389,9 @@ def mcrz(
         else:
             self.append(CRZGate(lam), control_qubits + [target_qubit])
     else:
+        if isinstance(lam, ParameterExpression):
+            raise QiskitError(f"Cannot synthesize MCRZ with unbound parameter: {lam}.")
+
```
The improved error messages look very helpful!

alexanderivrii commented on qiskit/circuit/library/standard_gates/rxx.py:
```diff
@@ -111,6 +114,39 @@ def _define(self):
+        num_ctrl_qubits: int = 1,
+        label: str | None = None,
+        ctrl_state: str | int | None = None,
+        annotated: bool | None = None,
+    ):
```
Would it have been possible to set the default value `annotated=True` without introducing `None`?

alexanderivrii reviewed (commented):
This looks really great!

One thing I am confused about is the new ternary semantics of `annotated`: with this PR it can be either `False`, `True` or `None`. Would it have been possible to achieve the same result without introducing the `None` value but by changing the default in the newly added control methods to `annotated=True`?

Previously the semantics of `annotated` was as follows: `False` - means do not create an annotated gate, this was required for backward compatibility. `True` - means introduce an annotated gate unless there is a friendlier gate available, e.g. `x.control(1, annotated=True)` would produce a `CX`-gate, not an annotated operation.

alexanderivrii commented:
Could you please also a test for reading/setting params for an annotated operation of an annotated operation of a gate?

Cryoris commented:
> Would it have been possible to achieve the same result without introducing the None value but by changing the default in the newly added control methods to annotated=True?

That would've been my preferred solution as well, but I thought the default argument cannot change in between classes derived from the same object. If that's allowed, then it would be the nicer solution 🙂

Edit: but the default value also changes with respect to the number of controls and whether the gate parameters are unbound... so we cannot really set it to `True` or `False` in the signature 🤔

alexanderivrii commented:
Suppose we keep the default `annotated=False` so as not change any previous behavior, but we modify the semantics of `gate.control(n, annotated)` to the following:
* if `annotated` is `False`, then we do not create an annotated operation, unless it depends on parameters
* if `annotated` is `True`, then we choose the version that is more convenient [as before]

In this way we do not change anything that used to work before, and obtain annotated operations for controlled parametric gates. Personally I prefer this slightly over the ternary semantics, but I am happy to change my opinion.

Cryoris commented:
The advantage of what you're describing is that we don't have to change the signature, which would be nice as it reduces the amount of required change. But this has two downsides:
1. We're implicitly changing default values (e.g. changing from `False` to `True` in some cases)
2. We're overriding user inputs (e.g. if a user explicitly sets `False` but we override it to `True`)

In the `None` scenario, I think it is easier to understand what default values are chosen, plus users can still force us to use whatever value they want. 

> if annotated is True, then we choose the version that is more convenient [as before]

I'm not sure I fully understand: if I do `RYGate(theta).control(annotated=True)` it still uses an `AnnotatedOperation` even though a `CRYGate` exists, which we could use.

alexanderivrii commented:
Thanks @Cryoris. I am not against merging this PR in the current form, and as it fixes an important set of problems we should definitely include it in 1.2. However, let me try to describe again the two options which I think we have.

First option (as I am proposing). We only have two values ``annotated=False`` and ``annotated=True``. I am thinking of these as user-specified _hints_ that the user passes to a gate's ``control``/``inverse``/``power`` methods, and Qiskit is allowed to override these values. For instance, it can override ``False`` to ``True`` for controlled parametric gates. And it can override ``True`` to ``False`` if Qiskit has a native representation of this gate that it can use. Personally, I think that `RYGate(theta).control(annotated=True)` (with 1 control) should be the `CRYGate`: since we have it in Qiskit, we should probably use it (in particular, maybe we have a great default definition for this gate). In this case, the default value is always ``False``.

Second option (as you are proposing). We have three values ``annotated=False``, ``annotated=True``, and ``annotated=None`` with the default value being ``None``. I am interpreting these as follows. The value ``None`` means let Qiskit choose the best possible representation. The value of ``False`` should then represent "do not create an ``AnnotatedOperation`` and fail horribly when writing ``RYGate(theta).control(annotated=False)``". The value of ``True`` means "always introduce an annotated operation, with in particular ``XGate().control(annotated=True)`` producing an ``AnnotatedOperation`` instead of ``CXGate()``."

After writing this down so explicitly, I actually start liking the second choice much better :). Especially if the default value is ``None`` in which case Qiskit decides how the gate should be implemented. And if a user wants to shoot oneself in the foot by providing non-default values, let it be. However, it also gives the user (and transpiler passes) much better control on what the gate really is. 

So, provided that we fully agree to go with the second option, the only thing I would like to ask you is to make the semantics of ``annotated`` clear both in the documentation and in the release notes. And also please check all the other places where ``annotated`` is used to make sure it adheres with this semantics.

Cryoris committed:
review comments

- use attribute error
- more clearly state the new None arg in reno and Gate class

Cryoris commented:
Option 2 it is then 🙂 I added more detailed explanations about the `None` argument and resolved the other comments 👍🏻

ElePT commented on qiskit/circuit/annotated_operation.py:
```diff
@@ -219,6 +220,29 @@ def power(self, exponent: float, annotated: bool = False):
+        if hasattr(self.base_op, "validate_parameter"):
+            return self.base_op.validate_parameter(parameter)
+
+        raise NotImplementedError(
```
```suggestion
        raise AttributeError(
```

Cryoris replied:
Done in 148b5c7

ElePT commented on qiskit/circuit/annotated_operation.py:
```diff
@@ -219,6 +220,29 @@ def power(self, exponent: float, annotated: bool = False):
+                f"Cannot set attribute ``params`` on the base operation {self.base_op}."
+            )
+
+    def validate_parameter(self, parameter: ParameterValueType) -> ParameterValueType:
```
Small question, would it be more coherent if we called it `validate_param`? I am a bit on the fence, I tend to prefer full words but I find it strange that the other methods say `params` and this one says `parameter`

Cryoris replied:
This is coming from `Instruction.validate_parameter`, to have `AnnotatedOperation` work as drop-in-replacement 🙂

ElePT commented on qiskit/circuit/library/standard_gates/rx.py:
```diff
@@ -90,15 +92,23 @@ def control(
-                as an annotated gate.
+                as an annotated gate. If ``None``, this is set to ``False``, unless the
+                gate contains free parameters and has more than one control qubit,
+                in which case it cannot yet be synthesized.
```
Is this trying to say that `None` will be treated as `True` if the gate contains free parameters and has more than one control? (I believe this is the case). Or that users are required to set it explicitly to `True`? If it's the first one, I would say:

```suggestion
                as an annotated gate. If ``None``, this is set to ``False`` except for 
                the case where the gate contains free parameters and has more than 
                one control qubit, where it is set to ``True``.
```

If it's the second one, I would say:

```suggestion
                as an annotated gate. If ``None``, this is set to ``False``. If the
                gate contains free parameters and has more than one control qubit,
                it cannot yet be synthesized unless this input is set to ``True``.
```

Cryoris replied:
I used the RXX message as you suggested below 🙂

ElePT commented on qiskit/circuit/library/standard_gates/rx.py:
```diff
@@ -90,15 +92,23 @@ def control(
             ctrl_state: control state expressed as integer,
                 string (e.g.``'110'``), or ``None``. If ``None``, use all 1s.
             annotated: indicates whether the controlled gate can be implemented
```
a bit out of scope, but does this indicate whether it can be implemented or whether it should be implemented? It looks to me like the second.

Cryoris replied:
I also think the second is correct... I can change it if we're already touching it 👍🏻

Cryoris replied:
Done in 7a03c4f

ElePT commented on qiskit/circuit/library/standard_gates/rxx.py:
```diff
@@ -111,6 +114,39 @@ def _define(self):
+            annotated: indicates whether the controlled gate can be implemented
+                as an annotated gate. If ``None``, this is set to ``True`` if
+                the gate contains free parameters, in which case it cannot
+                yet be synthesized.
```
This formulation already works a bit better than the one where I gave suggestions, they probably should be unified (I don't care about the specific message, it's just that I think it should be explicit about when it's set to true or false).

Cryoris replied:
Updated to use this one consistently in 7a03c4f 👍🏻

ElePT commented on releasenotes/notes/annotated-params-116288d5628f7ee8.yaml:
```diff
@@ -0,0 +1,23 @@
+---
+features_circuits:
+  - |
+    Add support for :meth:`.AnnotatedOperation.params`, which enables
```
should the reno include the other new method too?

```suggestion
    Added support for :meth:`.AnnotatedOperation.params` and :meth:`.AnnotatedOperation.validate_param`, which enable
```

ElePT commented on releasenotes/notes/annotated-params-116288d5628f7ee8.yaml:
```diff
@@ -0,0 +1,23 @@
+    annotated operations.
+fixes:
+  - |
+    Fix a series of issues when controlling parameterized standard gates.
```
This is optional but I think it makes the reno nicer :)

```suggestion
    Fixed a series of issues when controlling parameterized standard gates.
```

ElePT reviewed (commented):
I think the PR looks very good, just left a few minor wording suggestions.

Cryoris committed:
review from Elena

Cryoris committed:
Fix ``AttributeError`` test

Cryoris committed:
Merge branch 'main' into add-annotated-params

Cryoris committed:
lint

ElePT commented on releasenotes/notes/annotated-params-116288d5628f7ee8.yaml:
```diff
@@ -20,4 +21,10 @@ fixes:
+    ``None``, which allows Qiskit to choose whether to annotate a controlled operation.
+    Generally, the concrete implementation (``annotated=False``) is returned, if it
+    is available, otherwise we set ``annotated=True``. This allows, for example, to 
+    defer the synthesis of controlled, parameterized gates.
```
```suggestion
    The ``annotated`` argument of the :meth:`.Gate.control` method is now 
    ``None``, which allows Qiskit to choose whether to annotate a controlled operation.
    If the concrete implementation (``annotated=False``) is available, it will be returned by
    default. Otherwise, the annotated implementation will be returned (``annotated=True``). 
    This allows, for example, to defer the synthesis of controlled, parameterized gates.
```

ElePT commented on qiskit/circuit/gate.py:
```diff
@@ -118,7 +118,11 @@ def control(
+                if the controlled gate can directly be constructed, and otherwise
+                set to ``True``. This allows defering the construction process in case the
+                synthesis of the controlled gate requires more information (e.g.
+                values of unbound parameters).
```
```suggestion
                as an annotated gate. If ``None``, this is set to ``False``
                if the controlled gate can directly be constructed, and otherwise
                set to ``True``. This allows defering the construction process in case the
                synthesis of the controlled gate requires more information (e.g.
                values of unbound parameters).
```

ElePT reviewed (dismissed):
LGTM! I just added two comma-related minor suggestions. After they are applied I think the PR is ready to be merged.

Cryoris committed:
Apply suggestions from code review

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>
